Factor out functionality from update() that will be needed for layering

mkenigs opened the pull request:
Various refactors that will be helpful for sharing functionality between a legacy and layering update flow. Opening this against `master` per discussion in https://github.com/openshift/machine-config-operator/pull/2982#issuecomment-1059193703. I would like to get a CI run because I'm having trouble launching a cluster with code we're trying to get into `layering`

mkenigs commented on pkg/daemon/update.go:
```diff
@@ -114,9 +114,6 @@ func reloadService(name string) error {
 }
 
 // performPostConfigChangeAction takes action based on what postConfigChangeAction has been asked.
-// For non-reboot action, it applies configuration, updates node's config and state.
-// In the end uncordon node to schedule workload.
-// If at any point an error occurs, we reboot the node so that node has correct configuration.
```
Are these comments out of date? Or am I misunderstanding how error handling works here? It looks like it just returns errors

mkenigs commented:
/test e2e-gcp-op

mkenigs commented:
/test e2e-gcp-op

mkenigs commented:
/test e2e-gcp-op

cgwalters commented on pkg/daemon/drain.go:
```diff
@@ -153,15 +152,17 @@ func (dn *Daemon) performDrain() error {
 	return nil
 }
 
+type ReadFileFunc func(string) ([]byte, error)
```
This looks totally fine to me as is, but I'm thinking we may want to make this an interface in the future?

mkenigs replied:
Agreed. https://issues.redhat.com/browse/MCO-224

cgwalters commented on pkg/controller/common/constants.go:
```diff
@@ -1,6 +1,8 @@
 package common
 
 const (
+	MCONamespace = "openshift-machine-config-operator"
```
Probably worth converting at least one user in this commit.

cgwalters reviewed (approved):
Overall looks sane to me; I am a bit less confident in the later changes just because of the complexity of the code.  But, I think a lot of this is covered by our e2es.

cheesesashimi commented on pkg/daemon/drain_test.go:
```diff
@@ -359,9 +361,15 @@ location = "example.com/repo/test-img"
-			drain, err := isDrainRequired(test.actions, diffFileSet, oldIgnConfig, newIgnConfig)
-			if !reflect.DeepEqual(test.expectedAction, drain) {
-				t.Errorf("Failed determining drain behavior: expected: %v but result is: %v. Error: %v", test.expectedAction, drain, err)
+			readOldFile := func(path string) ([]byte, error) {
```
I noticed you duplicated these in a few places. A technique to reduce duplication is to have a function that returns a function (I know it sounds strange):

```golang
func getIgnitionFileDataReadFunc(ignConfig *ign3types.IgnConfig, path string) ReadFileFunc {
  return func(path string) ([]byte, error) {
    return ctrlcommon.GetIgnitionFileDataByPath(ignConfig, path)
  }
}
```

Then your calls become something like:

```golang
readOldFile := getIgnitionFileDataReadFunc(&oldIgnConfig, path)
readNewFile := getIgnitionFileDataReadFunc(&newIgnConfig, path)
```

mkenigs replied:
Ooo I like that more! Changed

cheesesashimi commented on pkg/daemon/update.go:
```diff
@@ -428,50 +428,58 @@ func (dn *Daemon) applyOSChanges(mcDiff machineConfigDiff, oldConfig, newConfig
 		"/etc/kubernetes/kubelet-ca.crt",
 		"/var/lib/kubelet/config.json",
 	}
+	if dn.os.IsFCOS() {
```
I was thinking about this as well.

cheesesashimi reviewed (commented):
I just have a couple minor questions and suggestions.

yuqi-zhang commented on pkg/daemon/update.go:
```diff
@@ -625,7 +622,16 @@ func (dn *Daemon) update(oldConfig, newConfig *mcfgv1.MachineConfig) (retErr err
+
+	// if dn.skipReboot, a reboot might be performed manually after update(), in which case we don't want to update state
+	// else reboot was not needed to apply the configuration, and we need to update state
+	if !dn.skipReboot {
```
I'm a bit confused at the use of skipReboot here. This is only for onceFrom right? Why is this specifically handled here?

mkenigs replied:
I missed adding this condition originally and it was causing bootstrap to fail. Looking at the old version of `performPostConfigChangeAction` it calls `dn.reboot` which I initially assumed would never return. But if `dn.skipReboot == true`, `dn.reboot` does return, and `performPostConfigChangeAction` returns early before updating state. So I had to add this to preserve that behavior. I can't say I fully understand why that's necessary because how state is managed is pretty unclear to me

yuqi-zhang reviewed (commented):
Some minor comments. I think it would be helpful for me at least if I can understand why some of these refactors are needed, for example:

1. Why does calculatePostConfigChangeAction need to be consolidated? 
2. What is the extra FCOS checks added for?
3. Why do we need to move isDrainRequired -> drainIfRequired

As a counter example, the explanation around readFileFunc is very nice

mkenigs commented on pkg/daemon/drain.go:
```diff
@@ -153,26 +152,40 @@ func (dn *Daemon) performDrain() error {
+func (dn *Daemon) drainIfRequired(actions, diffFileSet []string, readOldFile, readNewFile ReadFileFunc) (bool, error) {
+	// For any unhandled cases (although currently there aren't any), default to drain
+	drain := true
 	if ctrlcommon.InSlice(postConfigChangeActionReboot, actions) {
```
I don't think it's actually redundant because technically actions is a list (although currently it's only ever of length 1). So if we didn't have this if branch, if actions included both postConfigChangeActionReboot and postConfigChangeActionReloadCrio we could take eg the postConfigChangeActionReloadCrio branch and skip a drain

mkenigs commented on pkg/daemon/drain.go:
```diff
@@ -153,26 +152,40 @@ func (dn *Daemon) performDrain() error {
+
+// drainIfRequired determines whether node drain is required or not to apply config changes and performs the drain if required
+// returns whether drain was attempted
+func (dn *Daemon) drainIfRequired(actions, diffFileSet []string, readOldFile, readNewFile ReadFileFunc) (bool, error) {
```
Yep I can change it to that way. Is the advantage just less churn?

cheesesashimi replied:
Not just less churn but it also provides good separation of concerns. `isDrainRequired` answers the question of whether a drain is required or not. `drainIfRequired` can use that to ask the question and then perform the drain if it's needed.

mkenigs commented on pkg/daemon/drain_test.go:
```diff
@@ -348,6 +348,8 @@ location = "example.com/repo/test-img"
 	}
 
+	// kubeClient must be nil so that performDrain is a no-op
+	dn := Daemon{}
```
Removed because of one of @cheesesashimi 's suggestions

mkenigs commented:
>    1. Why does calculatePostConfigChangeAction need to be consolidated?

Added a much more detailed commit message (https://github.com/openshift/machine-config-operator/pull/2987/commits/c4b5dcf6143a895b9a278e366136303e59f77e83) Does that clear it up?

>     2. What is the extra FCOS checks added for?

The path for ssh keys is different on FCOS. Some of the explanation I added in the commit message should give some more context

>     3. Why do we need to move isDrainRequired -> drainIfRequired

Per @cheesesashimi 's feedback I just wrote a wrapper function. I'm just anticipating de-duplicating between legacy and layered update.

kikisdeliveryservice commented on pkg/daemon/drain.go:
```diff
@@ -176,6 +176,20 @@ func isDrainRequired(actions, diffFileSet []string, readOldFile, readNewFile Rea
 	return true, nil
 }
 
+func (dn *Daemon) drainIfRequired(actions, diffFileSet []string, readOldFile, readNewFile ReadFileFunc) error {
```
Can you explain what this is doing? Because I don't quite understand or know if this refactor should be pulled into master without accompanying code to make sense of it?

mkenigs replied:
It's just factoring code out of `update()` so it doesn't have to be duplicated when we want to perform drains for a layered update

kikisdeliveryservice replied:
I can't help but feel like that's should probably live with the other code that needs it? Feels weird to merge it if it's needed for something that isn't in master  or a clear optimization for master

cgwalters replied:
The counterbalance here is avoiding a later merge of layering into master being too large and painful, as well as the ongoing cost of rebases.

It's highly likely that there will be an intersection between layering and hypershift work for example.  And having prep patches merged into master helps avoid conflicts there too.

In this particular case, I think

- The code isn't *significantly* better or anything, but IMO it is a bit better
- The code isn't difficult to review
- CI covers all the functionality here

IMO, that on balance argues for a master merge.

kikisdeliveryservice commented on pkg/daemon/update.go:
```diff
@@ -486,6 +483,14 @@ func (dn *Daemon) update(oldConfig, newConfig *mcfgv1.MachineConfig) (retErr err
 		}
 	}
+	return nil
+}
```
similar comment to here - i dont really understand why this is necessary or how it fits into things?

kikisdeliveryservice commented on cmd/machine-config-operator/main.go:
```diff
@@ -8,10 +8,7 @@ import (
-	componentName      = "machine-config"
-	componentNamespace = "openshift-machine-config-operator"
-)
+const componentName = "machine-config"
```
I think the namespace constant changes could just be their own PR (8 file changes) and merge asap.

mkenigs commented:
Per @kikisdeliveryservice comments have opened:
https://github.com/openshift/machine-config-operator/pull/3020
https://github.com/openshift/machine-config-operator/pull/3021
https://github.com/openshift/machine-config-operator/pull/3022

mkenigs commented:
Will wait for other PRs to merge before revisiting commits left here

mkenigs commented:
@kikisdeliveryservice is this small enough now or should I split it again?

cgwalters commented:
Since this is targeting layering, I don't see any blockers to merge.
If there's stuff to address we can always do followups.
/lgtm

mkenigs commented:
/cherry-pick master
